### DISCUSSION TITLE
SDCard / SPI on Waveshare RP2040-LCD-1.28

### DISCUSSION THREAD

**Discussion by draeron** [earlephilhower/arduino-pico#2150] [General]
Trying to setup a SD card reader on a SPI bus on a RP2040-LCD-1.28 and for the life of me I can't get it to work (aka begin() return false). Tried with both default SPI0 layout and SPI1. I'm limited to the H1 header (on the right). Not totally sure if the screen use either of the SPI bus. Also tried changing the pins layout based on this comment. Also tried various SPI speed (full, half, 24h).

```/*
                 H1                                H2
            Pin#    Pin#                      Pin#    Pin#
    GPIO8   1       2    GPIO0        GND     1       2   GND
    GPIO9   3       4    GPIO1        VSYS    3       4   ADC_AVDD
    GPIO10  5       6    GPIO2        GPIO23  5       6   BOOT
    GPIO11  7       8    GPIO3        GPIO22  7       8   RUM
    GPIO12  9       10   GPIO4        GPIO21  9       10  GPIO29
    GPIO13  11      12   GPIO5        GPIO20  11      12  GPIO28
    GPIO14  13      14   GPIO6        GPIO19  13      14  GPIO27
    GPIO15  15      16   GPIO7        GPIO18  15      16  GPIO26
    SWCLK   17      18   VSYS         GPIO17  17      18  GPIO25
    SWDIP   19      20   GND          GPIO16  19      20  GPIO24


  The circuit:
    SD card attached to SPI bus as follows on RP2040:
   ************ SPI0 ************
   ** MISO (AKA RX) - pin 0, 4, or 16
   ** MOSI (AKA TX) - pin 3, 7, or 19,2
   ** CS            - pin 1, 5, or 17
   ** SCK           - pin 2, 6, or 18
   ************ SPI1 ************
   ** MISO (AKA RX) - pin  8 or 12
   ** MOSI (AKA TX) - pin 11 or 15
   ** CS            - pin  9 or 13
   ** SCK           - pin 10 or 14
*/
```

In the second comments, by pin we're talking GPIO IDs and not layout pins right ?

About the wiring, I saw some comment about needing to add pullup resistor but I'm pretty sure this is integrated on the board. There is also the question of if the MISO/MOSI label on the board should match (aka MOSI <-> MOSI vs MOSI <-> MISO) but i tried both and doesn't seems to make any difference.

<img src="https://github.com/earlephilhower/arduino-pico/assets/2256246/55e22187-8020-42cd-b648-ccbd2ceb9527" width="256">

Here is my code if someone can see anything obvious...

```c++
const int _MOSI = 7;  // AKA SPI TX
const int _SCK = 6;
const int _CS = 5;
const int _MISO = 4;  // AKA SPI RX
const int SD_SPI_SPEED = SPI_FULL_SPEED;

void setup() {
  Serial.begin(115200);
  while (!Serial) {
    yield();
  }

  Log.begin(LOG_LEVEL_VERBOSE, &Serial);

  sleep_ms(100);

  bool sdInitialized = false;

  // Initialize at the highest speed supported by the board that is
  // not over 50 MHz. Try a lower speed if SPI errors occur.
  while (!sdInitialized) {
    if (_MISO == 0 || _MISO == 4 || _MISO == 16) {
      SPI.setRX(_MISO);
      SPI.setTX(_MOSI);
      SPI.setSCK(_SCK);
      Log.noticeln(F("Using SPI0"));
      sdInitialized = SD.begin(_CS, SD_SPI_SPEED);
    } else if (_MISO == 8 || _MISO == 12) {
      SPI1.setRX(_MISO);
      SPI1.setTX(_MOSI);
      SPI1.setSCK(_SCK);
      Log.noticeln(F("Using SPI1"));
      sdInitialized = SD.begin(_CS, SD_SPI_SPEED, SPI1);
    } else {
      Log.fatalln(F("ERROR: Unknown SPI Configuration"));
      return;
    }

    if (!sdInitialized) {
      Log.errorln(F("initialization failed. Things to check:"));
      Log.errorln(F("* is a card inserted?"));
      Log.errorln(F("* is your wiring correct?"));
      Log.errorln(F("* did you change the chipSelect pin to match your shield or module?\n"));
      sleep_ms(1000);
    } else {
      Log.noticeln(F("SD initialization done"));
    }
  }

```

So if anyone gots any idea

**Comment by earlephilhower** [comment]
You photo of the adapter shows both a voltage regulator and a set of buffers.  So it may not be easily compatible with the Pico.  SD cards are 3.3V devices but a lot of those adapters were built for the old Arduino 5V boards so they need level shifters on the logic side and voltage regulators (and 5V in) on the input to power the card.

If you can find a board with no actives/passives at all that would be simplest.  I've even soldered wires to a full-size SD to microSD card pins and used that directly on the Pico, no trouble other than it looking ugly.  AX had some microSD boards which were just the socket pins wired to a header and they work great.

Also, MOSI@SD goes to MOSI@Pico, and MISO@SD goes to MISO@Pico.

**Comment by draeron** [comment]
now that you mention it. i've check the spec board... it take 5v input only.

```
- Level conversion circuit board that can interface level is 5V or 3.3V
- Power supply is 4.5V ~ 5.5V, 3.3V voltage regulator circuit board 
```
